File: aletheia/tools/rag_tool.py
"""
RAG Tool

Tool wrapper for RAG system.
Gọi aletheia.rag.HybridRetrieval để search documents.
"""

import logging
from typing import Any, Dict
from .base import Tool, ToolResult

logger = logging.getLogger(__name__)


class RAGTool(Tool):
    """
    Tool for document retrieval using RAG.

    Wraps aletheia.rag.HybridRetrieval to provide document search
    as a tool for the Agent.
    """

    def __init__(self):
        self._retrieval = None

    def _get_retrieval(self):
        """Lazy initialization of HybridRetrieval."""
        if self._retrieval is None:
            from aletheia.rag.retrieval.retrieval import HybridRetrieval

            self._retrieval = HybridRetrieval()
        return self._retrieval

    @property
    def name(self) -> str:
        return "document_search"

    @property
    def description(self) -> str:
        return (
            "Search through uploaded documents to find relevant information. "
            "Use this when the user asks questions about their documents, PDFs, or files. "
            "Returns relevant text passages with source citations."
        )

    @property
    def parameters_schema(self) -> Dict[str, Any]:
        return {
            "type": "object",
            "properties": {
                "query": {
                    "type": "string",
                    "description": "The search query about the documents",
                },
                "top_k": {
                    "type": "integer",
                    "description": "Number of results to return (default: 5)",
                    "default": 5,
                },
                "doc_id": {
                    "type": "string",
                    "description": "Optional: specific document ID to search within",
                    "default": None,
                },
            },
            "required": ["query"],
        }

    async def execute(self, arguments: Dict[str, Any]) -> ToolResult:
        """
        Execute document search.

        Args:
            arguments: Must contain 'query', optionally 'top_k' and 'doc_id'

        Returns:
            ToolResult with search results
        """
        try:
            query = arguments.get("query")
            logger.info(f"🔍 RAG Tool executing with query: '{query}'")
            if not query:
                return ToolResult(success=False, output="", error="Query is required")

            top_k = arguments.get("top_k", 5)
            doc_id = arguments.get("doc_id")

            # Call RAG
            retrieval = self._get_retrieval()

            # Use hybrid search - filter_doc_id là tên tham số đúng
            results = retrieval.hybrid_search(
                query=query, top_k=top_k, filter_doc_id=doc_id
            )
            # Log all results with truncation
            result_summary = []
            for i, result in enumerate(results, 1):
                text = result.get("text", "")
                metadata = result.get("metadata", {})
                filename = metadata.get("filename", "Unknown")
                page = metadata.get("page_num", "N/A")
                text_preview = text[:100] + "..." if len(text) > 100 else text
                result_summary.append(
                    f"  [{i}] {filename} (Page {page}): {text_preview}"
                )

            logger.info(
                f"📚 RAG search found {len(results)} results:\n"
                + "\n".join(result_summary)
            )

            # Format results
            if not results:
                return ToolResult(
                    success=True, output="No relevant documents found for this query."
                )

            # Build output with citations
            output_parts = []
            for i, result in enumerate(results, 1):
                text = result.get("text", "")
                metadata = result.get("metadata", {})
                filename = metadata.get("filename", "Unknown")
                page = metadata.get("page_num", "N/A")

                output_parts.append(f"[{i}] {filename} (Page {page}):\n{text}\n")

            return ToolResult(success=True, output="\n".join(output_parts))

        except Exception as e:
            return ToolResult(
                success=False, output="", error=f"Error searching documents: {str(e)}"
            )

File: src/tools/rag-tool.ts
import type { ToolResult } from "./base.js";
import { BaseTool, createToolResult } from "./base.js";
import type { HybridRetrieval, SearchResult } from "../rag/retrieval/retrieval.js";

export class RAGTool extends BaseTool {
  private _retrieval: HybridRetrieval | null = null;

  get name(): string {
    return "document_search";
  }

  get description(): string {
    return (
      "Search through uploaded documents to find relevant information. " +
      "Use this when the user asks questions about their documents, PDFs, or files. " +
      "Returns relevant text passages with source citations."
    );
  }

  get parametersSchema(): Record<string, unknown> {
    return {
      type: "object",
      properties: {
        query: {
          type: "string",
          description: "The search query about the documents",
        },
        top_k: {
          type: "integer",
          description: "Number of results to return (default: 5)",
          default: 5,
        },
        doc_id: {
          type: "string",
          description: "Optional: specific document ID to search within",
        },
      },
      required: ["query"],
    };
  }

  private async getRetrieval(): Promise<HybridRetrieval> {
    if (!this._retrieval) {
      const { HybridRetrieval } = await import("../rag/retrieval/retrieval.js");
      this._retrieval = new HybridRetrieval();
    }
    return this._retrieval;
  }

  async execute(args: Record<string, unknown>): Promise<ToolResult> {
    try {
      const query = args.query as string;
      if (!query) {
        return createToolResult(false, "", "Query is required");
      }

      const topK = (args.top_k as number) ?? 5;
      const docId = args.doc_id as string | undefined;

      console.log(`[rag-tool] Searching with query: '${query}'`);

      const retrieval = await this.getRetrieval();
      const results = await retrieval.hybridSearch(query, topK, 0.5, docId);

      if (results.length === 0) {
        return createToolResult(true, "No relevant documents found for this query.");
      }

      const outputParts: string[] = [];
      for (let i = 0; i < results.length; i++) {
        const result = results[i] as SearchResult;
        const text = (result.text as string) ?? "";
        const metadata = (result.metadata as Record<string, unknown>) ?? {};
        const filename = (metadata.filename as string) ?? "Unknown";
        const page = metadata.page_num ?? "N/A";
        outputParts.push(`[${i + 1}] ${filename} (Page ${page}):\n${text}\n`);
      }

      console.log(`[rag-tool] Found ${results.length} results`);
      return createToolResult(true, outputParts.join("\n"));
    } catch (e) {
      return createToolResult(
        false,
        "",
        `Error searching documents: ${e instanceof Error ? e.message : String(e)}`,
      );
    }
  }
}
